mautic/mautic
# Contact language aware content

Please answer the following questions. 

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Related user documentation PR URL | Todo
| Related developer documentation PR URL | Todo 
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | Y

# This PR has the following some PR branches as a part of it. Please merge the other two first, then allow me to back merge them into this one to ensure changes are not lost:

- [x] #2074 
- [x] #2130 - some parts were removed
- [x] #2073

### Required
#### Description:

This PR brings contact language awareness to emails, dynamic web content and improves the existing feature of landing pages. A new contact profile preferred locale field has been added. On page hits, this will automatically be populated with the browser's language preference. But can be customized in the contact's profile. 

Translations can be created for emails, landing pages, and dynamic web content. When a contact hits a page, or is sent an email or displayed a specific DWC item, Mautic will check to see if the contact has a preferred locale set, and if so then check to see if the item has a translation. If a match is found (or best match), it'll redirect, send, or display the translated content.

Email and landing page A/B tests can now have translations of those tests. 

It adds many trait classes and views to assist with future plugins that want to leverage these features. To be documented in dev docs. 

#### Steps to test this PR:
For each of these, be sure to clear cookies for accurate tests!

Landing pages:
1. Run doctrine:migrations:migrate
2. Create a landing page and choose a primary language. Make it apparent it's in one language. 
3. Clone the page and make the title/content apparent it's in a different language. Assign the page a different language and select that it is a translation of the first. 
4. Change your browser's preferred language to the translation (be sure it's listed first)
5. In an anonymous browser, browse to the public URL of the parent page. You should be redirected to the translation. Upon examining the contact's profile, it should have the locale preference set.

Emails:
1. Pick two contacts and give one a preferred language and the other something different. Add these two to a segment.
2. Repeat the same steps as 2-3 above only for a segment email.
3. Send the segment email.
4. Each lead should have received the version relative to their preferred locale.

DWC:
1. Repeat the same 2-3 steps above for a DWC item. 
2. Use a 3rd party website and embed the tracking pixel with email in the query (be sure email is set as publicly updatable). 
3. Embed the content for the DWC item. See Mautic's documentation.
4. In an anonymous browser, clear the cookies and visit the page. 
5. You should get the DWC item in the contact's preferred language.

#### List deprecations along with the new alternative:
1. Mautic.hideSpecificConfigFields() is now deprecated in favor of Mautic.initializeFormFieldVisibilitySwitcher() and is activated for all forms by default so that the data-hide-on and data-show-on attributes can be used Mautic wide.
